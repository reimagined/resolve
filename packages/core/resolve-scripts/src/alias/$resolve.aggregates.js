import {
  message,
  RESOURCE_ANY,
  RUNTIME_ENV_NOWHERE,
  IMPORT_INSTANCE
} from '../constants'
import importBabel from '../import_babel'
import { checkRuntimeEnv } from '../declare_runtime_env'
import importResource from '../import_resource'
import resolveFileOrModule from '../resolve_file_or_module'
import resolveFile from '../resolve_file'

export default ({ resolveConfig, isClient }) => {
  const imports = []
  const constants = []
  const exports = [``, `const aggregates = []`, ``]

  for (let index = 0; index < resolveConfig.aggregates.length; index++) {
    const aggregate = resolveConfig.aggregates[index]

    if (checkRuntimeEnv(aggregate.name)) {
      throw new Error(`${message.clientEnvError}.aggregates[${index}].name`)
    }
    constants.push(`const name_${index} = ${JSON.stringify(aggregate.name)}`)

    importResource({
      resourceName: `commands_${index}`,
      resourceValue: aggregate.commands,
      runtimeMode: RUNTIME_ENV_NOWHERE,
      importMode: RESOURCE_ANY,
      instanceMode: IMPORT_INSTANCE,
      imports: !isClient ? imports : [],
      constants: !isClient ? constants : []
    })

    if (isClient) {
      const clientCommands = Object.keys(
        aggregate.commands.constructor === String
          ? importBabel(resolveFile(aggregate.commands))
          : importBabel(resolveFileOrModule(aggregate.commands.module))(
              aggregate.commands.options,
              aggregate.commands.imports != null
                ? Object.keys(aggregate.commands.imports).reduce((acc, key) => {
                    acc[key] = importBabel(
                      resolveFile(aggregate.commands.imports[key])
                    )
                    return acc
                  }, {})
                : null
            )
      )

      constants.push(
        `const commands_${index} = {`,
        clientCommands
          .map(commandName => `  [\`${commandName}\`]: () => {}`)
          .join(',\r\n'),
        `}`
      )
    }

    exports.push(`aggregates.push({`)
    exports.push(`  name: name_${index}`)
    exports.push(`, commands: commands_${index}`)

    if (!isClient) {
      importResource({
        resourceName: `serializeState_${index}`,
        resourceValue: aggregate.serializeState,
        runtimeMode: RUNTIME_ENV_NOWHERE,
        importMode: RESOURCE_ANY,
        instanceMode: IMPORT_INSTANCE,
        instanceFallback:
          'resolve-runtime/lib/common/defaults/json-serialize-state.js',
        imports,
        constants
      })

      exports.push(`, serializeState: serializeState_${index}`)

      importResource({
        resourceName: `deserializeState_${index}`,
        resourceValue: aggregate.deserializeState,
        runtimeMode: RUNTIME_ENV_NOWHERE,
        importMode: RESOURCE_ANY,
        instanceMode: IMPORT_INSTANCE,
        instanceFallback:
          'resolve-runtime/lib/common/defaults/json-deserialize-state.js',
        imports,
        constants
      })

      exports.push(`, deserializeState: deserializeState_${index}`)

      if (aggregate.projection != null) {
        importResource({
          resourceName: `projection_${index}`,
          resourceValue: aggregate.projection,
          runtimeMode: RUNTIME_ENV_NOWHERE,
          importMode: RESOURCE_ANY,
          instanceMode: IMPORT_INSTANCE,
          calculateHash: 'resolve-aggregate-projection-hash',
          imports,
          constants
        })

        exports.push(`, projection: projection_${index}`)
        exports.push(`, invariantHash: projection_${index}_hash`)
      }
    }

    exports.push(`})`, ``)
  }

  exports.push(`export default aggregates`)

  return {
    code: [...imports, ...constants, ...exports].join('\r\n')
  }
}
