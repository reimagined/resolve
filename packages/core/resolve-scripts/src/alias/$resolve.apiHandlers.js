import {
  message,
  RESOURCE_ANY,
  RUNTIME_ENV_ANYWHERE,
  IMPORT_INSTANCE
} from '../constants'
import { checkRuntimeEnv } from '../declare_runtime_env'
import importResource from '../import_resource'

export default ({ resolveConfig, isClient }) => {
  if (isClient) {
    throw new Error(
      `${message.serverAliasInClientCodeError}$resolve.apiHandlers`
    )
  }

  if (!resolveConfig.apiHandlers) {
    throw new Error(`${message.configNotContainSectionError}.apiHandlers`)
  }

  if (checkRuntimeEnv(resolveConfig.apiHandlers)) {
    throw new Error(`${message.clientEnvError}.apiHandlers`)
  }

  const imports = [``]
  const constants = [``]
  const exports = [``, `const apiHandlers = []`, ``]

  for (let index = 0; index < resolveConfig.apiHandlers.length; index++) {
    const apiHandler = resolveConfig.apiHandlers[index]

    if (checkRuntimeEnv(apiHandler.path)) {
      throw new Error(`${message.clientEnvError}.apiHandlers[${index}].path`)
    }
    constants.push(`const path_${index} = ${JSON.stringify(apiHandler.path)}`)

    if (checkRuntimeEnv(apiHandler.method)) {
      throw new Error(`${message.clientEnvError}.apiHandlers[${index}].method`)
    }
    constants.push(
      `const method_${index} = ${JSON.stringify(apiHandler.method)}`
    )

    importResource({
      resourceName: `controller_${index}`,
      resourceValue: apiHandler.controller,
      runtimeMode: RUNTIME_ENV_ANYWHERE,
      importMode: RESOURCE_ANY,
      instanceMode: IMPORT_INSTANCE,
      imports,
      constants
    })

    exports.push(`apiHandlers.push({`, `  path: path_${index}`)
    exports.push(`, controller: controller_${index}`)
    exports.push(`, method: method_${index}`)
    exports.push(`})`, ``)
  }

  exports.push(`export default apiHandlers`)

  return {
    code: [...imports, ...constants, ...exports].join('\r\n')
  }
}
