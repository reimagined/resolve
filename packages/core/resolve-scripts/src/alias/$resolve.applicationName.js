import resolveFile from '../resolve_file'

export default () => {
  const { name: applicationName } = require(resolveFile('package.json'))
  const exports = []

  exports.push(
    `const applicationName = ${JSON.stringify(applicationName)}`,
    ``,
    `export default applicationName`
  )

  return exports.join('\r\n')
}
