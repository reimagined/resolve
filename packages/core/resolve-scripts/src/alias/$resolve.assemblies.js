import fs from 'fs'
import path from 'path'

import { includeAlias as excludeConstantAlias } from './$resolve.constants'

const excludeAlias = [
  'clientEntry',
  'hotModuleReplacement',
  'localEntry',
  'cloudEntry',
  'apiHandlers',
  'aggregates',
  'readModels',
  'viewModels',
  'sagas',
  'assemblies',
  ...excludeConstantAlias
]

export default () => {
  const exports = []

  const alias = fs
    .readdirSync(__dirname)
    .filter(filename => path.extname(filename) === '.js')
    .map(filename => path.basename(filename, '.js').replace('$resolve.', ''))
    .filter(alias => !excludeAlias.includes(alias))

  for (const name of alias) {
    exports.push(`import ${name} from '$resolve.${name}'`)
  }

  exports.push(``, `export default {`, ` ${alias.join(',\r\n')}`, `}`)

  return {
    code: exports.join('\r\n')
  }
}
