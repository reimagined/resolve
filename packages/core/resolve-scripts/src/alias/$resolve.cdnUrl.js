import declareRuntimeEnv, { injectRuntimeEnv } from '../declare_runtime_env'

export default ({ resolveConfig, isClient }) => {
  const exports = []
  if (resolveConfig.hasOwnProperty('uploadAdapter')) {
    exports.push(
      `const cdnUrl = ${injectRuntimeEnv(
        declareRuntimeEnv('RESOLVE_UPLOADER_CDN_URL'),
        isClient
      )}`
    )
  } else {
    exports.push(`const cdnUrl = null`)
  }

  exports.push(`export default cdnUrl`)

  return exports.join('\r\n')
}
