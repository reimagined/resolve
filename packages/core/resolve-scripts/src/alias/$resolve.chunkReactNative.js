export default () => {
  const exports = []

  exports.push(
    `export { default as aggregates } from '$resolve.aggregates'`,
    `export { default as aggregateActions } from '$resolve.aggregateActions'`,
    `export { default as viewModels } from '$resolve.viewModels'`,
    `export { default as readModels } from '$resolve.readModels'`,
    `export { default as rootPath } from '$resolve.rootPath'`,
    `export { default as staticPath } from '$resolve.staticPath'`,
    `export { default as jwtCookie } from '$resolve.jwtCookie'`,
    `export { default as applicationName } from '$resolve.applicationName'`,
    `export { default as subscribeAdapter } from '$resolve.subscribeAdapter'`,
    `export { default as customConstants } from '$resolve.customConstants'`
  )

  return {
    code: exports.join('\r\n')
  }
}
