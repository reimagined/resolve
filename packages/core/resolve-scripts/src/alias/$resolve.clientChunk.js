import getClientGlobalEnvObject from '../client_global_object'
import { checkRuntimeEnv } from '../declare_runtime_env'
import { message } from '../constants'

export default ({ resolveConfig, isClient }) => {
  if (!isClient) {
    throw new Error(`${message.clientAliasInServerCodeError}.clientAssemblies`)
  }
  const exports = []

  exports.push(
    `import interopRequireDefault from "@babel/runtime/helpers/interopRequireDefault"`,
    `const clientGlobalObject = ${getClientGlobalEnvObject()}`,
    `const defaultRuntimeEnv = {}`
  )

  const clientEnvs = []
  void JSON.stringify(resolveConfig, (key, value) => {
    if (checkRuntimeEnv(value)) {
      clientEnvs.push(value)
    }
    return value
  })

  for (const clientEnv of clientEnvs) {
    if (process.env[String(clientEnv)] != null) {
      exports.push(`Object.defineProperty(
        defaultRuntimeEnv,
        ${JSON.stringify(String(clientEnv))},
        {
          enumerable: true,
          value: ${JSON.stringify(process.env[String(clientEnv)])}
        }
      )`)
    } else {
      exports.push(`Object.defineProperty(
        defaultRuntimeEnv,
        ${JSON.stringify(String(clientEnv))},
        {
          enumerable: true,
          value: ${JSON.stringify(String(clientEnv.defaultValue))}
        }
      )`)
    }
  }

  exports.push(`
    if(clientGlobalObject.__RESOLVE_RUNTIME_ENV__ == null) {
      clientGlobalObject.__RESOLVE_RUNTIME_ENV__ = defaultRuntimeEnv
      try {
        if(window == null || clientGlobalObject !== window) {
          throw new Error('Skip warning')
        }
        console.warn(\`
          Client-runtime variables have been set to default values since __RESOLVE_RUNTIME_ENV__ is not defined
        \`)
      } catch(e) {}
    }
  `)

  exports.push(
    `const clientChunk = {`,
    `  clientImports: interopRequireDefault(require('$resolve.clientImports')).default,`,
    `  viewModels: interopRequireDefault(require('$resolve.viewModels')).default,`,
    `  rootPath: interopRequireDefault(require('$resolve.rootPath')).default,`,
    `  staticPath: interopRequireDefault(require('$resolve.staticPath')).default,`,
    `  jwtCookie: interopRequireDefault(require('$resolve.jwtCookie')).default,`,
    `  applicationName: interopRequireDefault(require('$resolve.applicationName')).default,`,
    `  subscriber: interopRequireDefault(require('resolve-client/lib/subscribe-adapter')).default,`,
    `  customConstants: interopRequireDefault(require('$resolve.customConstants')).default`,
    `}`
  )

  exports.push(`export default clientChunk`)

  return exports.join('\r\n')
}
