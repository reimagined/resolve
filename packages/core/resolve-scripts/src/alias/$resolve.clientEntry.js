import loaderUtils from 'loader-utils'
import {
  RUNTIME_ENV_NOWHERE,
  RESOURCE_INSTANCE_ONLY,
  IMPORT_INSTANCE
} from '../constants'
import importResource from '../import_resource'

export default ({ resolveConfig, isClient }, resourceQuery) => {
  if (!/^\?/.test(resourceQuery)) {
    throw new Error(
      `Resource $resolve.clientEntry should be retrieved with resource query`
    )
  }

  const inputFile = loaderUtils.parseQuery(resourceQuery).inputFile
  const clientEntry = resolveConfig.clientEntries.find(
    entry =>
      `${inputFile}` === (Array.isArray(entry) ? `${entry[0]}` : `${entry}`)
  )

  if (clientEntry == null) {
    throw new Error(`Resource $resolve.clientEntry "${inputFile}" not found`)
  }

  const moduleType = Array.isArray(clientEntry)
    ? clientEntry[1].moduleType
    : 'iife'

  const imports = []
  const constants = []
  const exports = []

  if (isClient) {
    imports.push(`import clientImports from '$resolve.clientImports'`)
    imports.push(`import clientChunk from '$resolve.clientChunk'`)
    imports.push(`import localS3Constants from '$resolve.localS3Constants'`)
    constants.push(`const entryArgs = {
      clientImports,
      localS3Constants,
      ...clientChunk
    }`)
  } else {
    imports.push(`import serverImports from '$resolve.serverImports'`)
    imports.push(`import seedClientEnvs from '$resolve.seedClientEnvs'`)
    imports.push(`import constants from '$resolve.constants'`)
    imports.push(`import eventListeners from '$resolve.eventListeners'`)
    imports.push(`import aggregates from '$resolve.aggregates'`)
    imports.push(`import readModels from '$resolve.readModels'`)
    imports.push(`import viewModels from '$resolve.viewModels'`)
    imports.push(`import sagas from '$resolve.sagas'`)
    imports.push(`import schedulers from '$resolve.schedulers'`)
    imports.push(`import localS3Constants from '$resolve.localS3Constants'`)
    constants.push(`const entryArgs = {
      serverImports,
      seedClientEnvs,
      constants,
      eventListeners,
      aggregates,
      readModels,
      viewModels,
      sagas,
      schedulers,
      localS3Constants
    }`)
  }

  importResource({
    resourceName: `clientEntry`,
    resourceValue: inputFile,
    runtimeMode: RUNTIME_ENV_NOWHERE,
    importMode: RESOURCE_INSTANCE_ONLY,
    instanceMode: IMPORT_INSTANCE,
    imports,
    constants
  })

  if (moduleType === 'iife') {
    exports.push(`clientEntry(entryArgs)`)
  } else {
    exports.push(`const boundClientEntry = clientEntry.bind(null, entryArgs)`)
    exports.push(`export default boundClientEntry`)
  }

  return [...imports, ...constants, ...exports].join('\r\n')
}
