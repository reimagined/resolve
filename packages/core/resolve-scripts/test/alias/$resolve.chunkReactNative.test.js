import alias from '../../src/alias/$resolve.chunkReactNative'
import normalizePaths from './normalize_paths'

test('works correctly', () => {
  expect(normalizePaths('\r\n' + alias({}) + '\r\n')).toMatchSnapshot()
})
