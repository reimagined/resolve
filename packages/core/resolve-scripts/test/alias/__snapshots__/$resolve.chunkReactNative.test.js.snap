// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works correctly 1`] = `
"
import interopRequireDefault from \\"@babel/runtime/helpers/interopRequireDefault\\"
const clientGlobalObject = ((() => {
    const globalObject = [() => window, () => global, () => self].reduce(
      (acc, recognizer) => {
        try {
          return acc != null ? acc : recognizer()
        } catch(e) {}
      },
      null
    )
    
    if (globalObject == null) {
      throw new Error('Client global object recognition failed')
    }

    if(true) {
      return globalObject
    }

    const key = undefined
    if(globalObject[key] == null) {
      throw new Error(\`Client global key \${key} object recognition failed\`)
    }
  
    return globalObject[key]
  })())
clientGlobalObject.__RESOLVE_RUNTIME_ENV__ = { }
import { createActions } from 'resolve-redux'
export const viewModels = interopRequireDefault(require('$resolve.viewModels')).default
export const rootPath = interopRequireDefault(require('$resolve.rootPath')).default
export const staticPath = interopRequireDefault(require('$resolve.staticPath')).default
export const jwtCookie = interopRequireDefault(require('$resolve.jwtCookie')).default
export const applicationName = interopRequireDefault(require('$resolve.applicationName')).default
export const subscriber = interopRequireDefault(require('resolve-client/lib/subscribe-adapter')).default
export const customConstants = interopRequireDefault(require('$resolve.customConstants')).default
"
`;
